content/blog/2021-07-tonic-0-5.md

```diff
@@ -0,0 +1,174 @@
+
+```rust
+use tonic::transport::Server;
+use tower_http::{
+    auth::RequireAuthorizationLayer,
+    trace::TraceLayer,
+};
+
+// The stack of middleware our service will be wrapped in
+let layer = tower::ServiceBuilder::new()
+    // High level tracing of requests and responses
+    .layer(TraceLayer::new_for_grpc())
+    // Authorize all requests using a token
+    .layer(RequireAuthorizationLayer::bearer("my-secret-token"))
+    // Convert our `ServiceBuilder` into a `tower::Layer`
+    .into_inner();
+
+Server::builder()
+    // Apply our middleware stack to the server
+    .layer(layer)
+    .add_service(GreeterServer::new(MyGreeter)
+    .serve(addr)
+    .await?;
+```
+
+# More flexible interceptors
+
+Interceptors are lightweight middleware that can, among other things, be used to
+modify the meta data of incoming requests and optionally reject them with a
+status.
+
+Interceptors have always been fairly limited however. For example, you couldn't
+combine multiple interceptors and would instead have to use Tower's [`Service`]
+abstraction, which is more powerful but also requires more code to setup.
+
+For 0.5 we have a new interceptor API that is as easy to use as before but now
+uses Tower internally. That means interceptors can be applied in the same way as
+any other Tower middleware. For example adding multiple interceptors to a client
+is now possible with:
+
+```rust
+use tonic::{
+  Request, Status,
+  service::interceptor_fn,
+  transport::Endpoint
+};
+
+fn intercept_one(req: Request<()>) -> Result<Request<()>, Status> {
+    // ...
+}
+
+fn intercept_two(req: Request<()>) -> Result<Request<()>, Status> {
+    // ...
+}
+
+let channel = Endpoint::from_static("http://[::1]:50051").connect_lazy()?;
+
+let intercepted_channel = tower::ServiceBuilder::new()
+    .layer(interceptor_fn(intercept_one))
+    .layer(interceptor_fn(intercept_two))
+    .service(channel);
+
+let client = GreeterClient::new(intercepted_channel);
+```
+
+0.5 includes many other smaller features and improvements. The [changelog] has
+all the details.
+
+As always, if you have questions you can find us in `#tonic` in the [Tokio
+Discord] server.
+
+<div style="text-align:right">&mdash; David Pedersen (<a href="https://github.com/davidpdrsn">@davidpdrsn</a>)</div>
+
+[Tonic]: https://crates.io/crates/tonic
+[grpc-web]: https://github.com/grpc/grpc-web
+[tonic-web]: https://crates.io/crates/tonic-web
+[compression-docs]: https://docs.rs/tonic/0.5.0/tonic/index.html?search=gzip
+[Tower]: https://crates.io/crates/tower
+[tower-http]: https://crates.io/crates/tower-http
+[`Service`]: https://docs.rs/tower/latest/tower/trait.Service.html
+[changelog]: https://github.com/hyperium/tonic/blob/master/CHANGELOG.md
+[Tokio Discord]: https://discord.gg/tokio
+[`Server::layer`]: https://docs.rs/tonic/latest/tonic/transport/server/struct.Server.html#method.layer
+[tonic-build]: https://crates.io/crates/tonic-build
```

Comment:
if you make the changes i suggested  to add code formatting to crate names, you'll also want to do this:
```suggestion
[grpc-web]: https://github.com/grpc/grpc-web
[`tonic-web`]: https://crates.io/crates/tonic-web
[compression-docs]: https://docs.rs/tonic/0.5.0/tonic/index.html?search=gzip
[Tower]: https://crates.io/crates/tower
[`tower-http`]: https://crates.io/crates/tower-http
[`Service`]: https://docs.rs/tower/latest/tower/trait.Service.html
[changelog]: https://github.com/hyperium/tonic/blob/master/CHANGELOG.md
[Tokio Discord]: https://discord.gg/tokio
[`Server::layer`]: https://docs.rs/tonic/latest/tonic/transport/server/struct.Server.html#method.layer
[`tonic-build`]: https://crates.io/crates/tonic-build
```